File: auth-profile-nestjs/auth/src/modules/identity/services/token.service.ts
import {
  Injectable,
  UnauthorizedException,
  Logger,
  InternalServerErrorException,
} from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import { ConfigService } from '@nestjs/config';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository, LessThan } from 'typeorm';
import * as bcrypt from 'bcrypt';
import { v4 as uuidv4 } from 'uuid';

import { RefreshToken } from '../entities';
import {
  AccessTokenPayload,
  RefreshTokenPayload,
  AuthTokens,
  DeviceInfo,
} from '../interfaces';
import { TOKEN_CONFIG, ERROR_MESSAGES } from '../constants';
import { RedisCacheService } from './redis-cache.service';

@Injectable()
export class TokenService {
  private readonly logger = new Logger(TokenService.name);

  constructor(
    @InjectRepository(RefreshToken)
    private readonly refreshTokenRepository: Repository<RefreshToken>,
    private readonly jwtService: JwtService,
    private readonly configService: ConfigService,
    private readonly redisCacheService: RedisCacheService,
  ) {}

  /**
   * Generate access and refresh token pair
   */
  async generateTokenPair(
    userId: string,
    email: string,
    deviceInfo: DeviceInfo,
  ): Promise<AuthTokens> {
    try {
      const accessTokenJti = uuidv4();
      const refreshTokenJti = uuidv4();

      const accessTokenPayload: AccessTokenPayload = {
        sub: userId,
        email,
        jti: accessTokenJti,
      };

      const refreshTokenPayload: RefreshTokenPayload = {
        sub: userId,
        jti: refreshTokenJti,
        deviceId: deviceInfo.deviceId,
      };

      const [accessToken, refreshToken] = await Promise.all([
        this.jwtService.signAsync(accessTokenPayload, {
          secret: this.configService.getOrThrow<string>('JWT_ACCESS_SECRET'),
          expiresIn: TOKEN_CONFIG.ACCESS_TOKEN_EXPIRY,
        }),
        this.jwtService.signAsync(refreshTokenPayload, {
          secret: this.configService.getOrThrow<string>('JWT_REFRESH_SECRET'),
          expiresIn: TOKEN_CONFIG.REFRESH_TOKEN_EXPIRY,
        }),
      ]);

      // Save refresh token to database
      await this.saveRefreshToken(
        refreshTokenJti,
        userId,
        refreshToken,
        deviceInfo,
      );

      // Calculate expiresIn in seconds
      const decoded = this.jwtService.decode(accessToken) as AccessTokenPayload;
      const expiresIn = decoded.exp ? decoded.exp - Math.floor(Date.now() / 1000) : 900;

      return {
        accessToken,
        refreshToken,
        expiresIn,
      };
    } catch (error) {
      this.logger.error(
        'Failed to generate token pair',
        error instanceof Error ? error.stack : String(error),
      );
      throw new InternalServerErrorException('Failed to generate tokens');
    }
  }

  /**
   * Verify and decode access token
   */
  async verifyAccessToken(token: string): Promise<AccessTokenPayload> {
    try {
      const payload = await this.jwtService.verifyAsync<AccessTokenPayload>(token, {
        secret: this.configService.getOrThrow<string>('JWT_ACCESS_SECRET'),
      });

      // Check if token is blacklisted
      const isBlacklisted = await this.redisCacheService.isTokenBlacklisted(payload.jti);
      if (isBlacklisted) {
        throw new UnauthorizedException(ERROR_MESSAGES.TOKEN_REVOKED);
      }

      return payload;
    } catch (error) {
      if (error instanceof UnauthorizedException) {
        throw error;
      }
      this.logger.debug('Access token verification failed');
      throw new UnauthorizedException(ERROR_MESSAGES.TOKEN_INVALID);
    }
  }

  /**
   * Verify and decode refresh token
   */
  async verifyRefreshToken(token: string): Promise<RefreshTokenPayload> {
    try {
      return await this.jwtService.verifyAsync<RefreshTokenPayload>(token, {
        secret: this.configService.getOrThrow<string>('JWT_REFRESH_SECRET'),
      });
    } catch {
      throw new UnauthorizedException(ERROR_MESSAGES.REFRESH_TOKEN_INVALID);
    }
  }

  /**
   * Rotate refresh token (revoke old, issue new)
   */
  async rotateRefreshToken(
    oldToken: string,
    deviceInfo: DeviceInfo,
  ): Promise<{ userId: string; email: string; tokens: AuthTokens }> {
    try {
      // Verify the old refresh token
      const payload = await this.verifyRefreshToken(oldToken);

      // Find and validate stored token
      const storedToken = await this.refreshTokenRepository.findOne({
        where: {
          id: payload.jti,
          userId: payload.sub,
          isRevoked: false,
        },
        relations: ['user'],
      });

      if (!storedToken) {
        throw new UnauthorizedException(ERROR_MESSAGES.TOKEN_REVOKED);
      }

      if (new Date() > storedToken.expiresAt) {
        throw new UnauthorizedException(ERROR_MESSAGES.REFRESH_TOKEN_EXPIRED);
      }

      // Revoke the old token
      storedToken.isRevoked = true;
      await this.refreshTokenRepository.save(storedToken);

      // Generate new token pair
      const tokens = await this.generateTokenPair(
        storedToken.userId,
        storedToken.user.email,
        deviceInfo,
      );

      return {
        userId: storedToken.userId,
        email: storedToken.user.email,
        tokens,
      };
    } catch (error) {
      if (error instanceof UnauthorizedException) {
        throw error;
      }
      this.logger.error(
        'Token rotation failed',
        error instanceof Error ? error.stack : String(error),
      );
      throw new UnauthorizedException(ERROR_MESSAGES.REFRESH_TOKEN_INVALID);
    }
  }

  /**
   * Revoke a specific token by JTI
   */
  async revokeToken(jti: string, remainingTtlSeconds: number): Promise<void> {
    try {
      if (remainingTtlSeconds > 0) {
        await this.redisCacheService.blacklistToken(jti, remainingTtlSeconds);
      }
    } catch (error) {
      this.logger.error(
        `Failed to revoke token ${jti}`,
        error instanceof Error ? error.stack : String(error),
      );
    }
  }

  /**
   * Revoke all refresh tokens for a user
   */
  async revokeAllUserTokens(userId: string): Promise<void> {
    try {
      await this.refreshTokenRepository.update(
        { userId, isRevoked: false },
        { isRevoked: true },
      );
      this.logger.debug(`Revoked all tokens for user ${userId}`);
    } catch (error) {
      this.logger.error(
        `Failed to revoke all tokens for user ${userId}`,
        error instanceof Error ? error.stack : String(error),
      );
    }
  }

  /**
   * Revoke refresh tokens for a specific device
   */
  async revokeDeviceTokens(userId: string, deviceId: string): Promise<void> {
    try {
      await this.refreshTokenRepository.update(
        { userId, deviceInfo: deviceId, isRevoked: false },
        { isRevoked: true },
      );
      this.logger.debug(`Revoked tokens for user ${userId}, device ${deviceId}`);
    } catch (error) {
      this.logger.error(
        `Failed to revoke device tokens for user ${userId}`,
        error instanceof Error ? error.stack : String(error),
      );
    }
  }

  /**
   * Clean up expired tokens (scheduled task)
   */
  async cleanupExpiredTokens(): Promise<number> {
    try {
      const result = await this.refreshTokenRepository.delete({
        expiresAt: LessThan(new Date()),
      });
      const deletedCount = result.affected ?? 0;
      this.logger.log(`Cleaned up ${deletedCount} expired refresh tokens`);
      return deletedCount;
    } catch (error) {
      this.logger.error(
        'Failed to cleanup expired tokens',
        error instanceof Error ? error.stack : String(error),
      );
      return 0;
    }
  }

  /**
   * Get active sessions for a user
   */
  async getUserSessions(userId: string): Promise<RefreshToken[]> {
    return this.refreshTokenRepository.find({
      where: { userId, isRevoked: false },
      order: { createdAt: 'DESC' },
    });
  }

  /**
   * Save refresh token hash to database
   */
  private async saveRefreshToken(
    jti: string,
    userId: string,
    token: string,
    deviceInfo: DeviceInfo,
  ): Promise<RefreshToken> {
    const tokenHash = await bcrypt.hash(token, TOKEN_CONFIG.BCRYPT_SALT_ROUNDS);

    const expiresAt = new Date();
    expiresAt.setDate(expiresAt.getDate() + TOKEN_CONFIG.REFRESH_TOKEN_EXPIRY_DAYS);

    const refreshTokenEntity = this.refreshTokenRepository.create({
      id: jti,
      userId,
      tokenHash,
      deviceInfo: deviceInfo.deviceId,
      ipAddress: deviceInfo.ipAddress,
      expiresAt,
      isRevoked: false,
    });

    return this.refreshTokenRepository.save(refreshTokenEntity);
  }

  /**
   * Decode token without verification (for reading claims)
   */
  decodeToken<T extends AccessTokenPayload | RefreshTokenPayload>(token: string): T {
    return this.jwtService.decode(token) as T;
  }
}

File: auth/internal/service/token_service.go
package service

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"time"

	"github.com/golang-jwt/jwt/v5"
	"github.com/google/uuid"
	"golang.org/x/crypto/bcrypt"

	"microservices/auth/internal/cache"
	"microservices/auth/internal/config"
	"microservices/auth/internal/model"
	"microservices/auth/internal/repository"
	"microservices/auth/pkg/errors"
	"microservices/auth/pkg/logger"
)

const (
	BcryptCost = 12
)

type AccessTokenClaims struct {
	jwt.RegisteredClaims
	Email string `json:"email"`
}

type RefreshTokenClaims struct {
	jwt.RegisteredClaims
	DeviceID string `json:"deviceId"`
}

type AuthTokens struct {
	AccessToken  string `json:"accessToken"`
	RefreshToken string `json:"refreshToken"`
	ExpiresIn    int64  `json:"expiresIn"`
}

type DeviceInfo struct {
	DeviceID  string
	IPAddress string
	UserAgent string
}

type TokenService struct {
	cfg              *config.Config
	refreshTokenRepo repository.RefreshTokenRepository
	cache            *cache.RedisClient
}

func NewTokenService(
	cfg *config.Config,
	refreshTokenRepo repository.RefreshTokenRepository,
	cache *cache.RedisClient,
) *TokenService {
	return &TokenService{
		cfg:              cfg,
		refreshTokenRepo: refreshTokenRepo,
		cache:            cache,
	}
}

func (s *TokenService) GenerateTokenPair(ctx context.Context, userID uuid.UUID, email string, deviceInfo *DeviceInfo) (*AuthTokens, error) {
	jti := uuid.New().String()
	now := time.Now()

	// Generate access token
	accessClaims := AccessTokenClaims{
		RegisteredClaims: jwt.RegisteredClaims{
			Subject:   userID.String(),
			ID:        jti,
			IssuedAt:  jwt.NewNumericDate(now),
			ExpiresAt: jwt.NewNumericDate(now.Add(s.cfg.JWT.AccessExpiry)),
		},
		Email: email,
	}

	accessToken := jwt.NewWithClaims(jwt.SigningMethodHS256, accessClaims)
	accessTokenString, err := accessToken.SignedString([]byte(s.cfg.JWT.AccessSecret))
	if err != nil {
		return nil, errors.Wrap(err, "TOKEN_GENERATION_FAILED", "Failed to generate access token", 500)
	}

	// Generate refresh token
	refreshJti := uuid.New().String()
	refreshClaims := RefreshTokenClaims{
		RegisteredClaims: jwt.RegisteredClaims{
			Subject:   userID.String(),
			ID:        refreshJti,
			IssuedAt:  jwt.NewNumericDate(now),
			ExpiresAt: jwt.NewNumericDate(now.Add(s.cfg.JWT.RefreshExpiry)),
		},
		DeviceID: deviceInfo.DeviceID,
	}

	refreshToken := jwt.NewWithClaims(jwt.SigningMethodHS256, refreshClaims)
	refreshTokenString, err := refreshToken.SignedString([]byte(s.cfg.JWT.RefreshSecret))
	if err != nil {
		return nil, errors.Wrap(err, "TOKEN_GENERATION_FAILED", "Failed to generate refresh token", 500)
	}

	// Hash refresh token for storage
	tokenHash := hashToken(refreshTokenString)

	// Save refresh token to database
	refreshTokenEntity := &model.RefreshToken{
		ID:        uuid.MustParse(refreshJti),
		UserID:    userID,
		TokenHash: tokenHash,
		DeviceID:  &deviceInfo.DeviceID,
		IPAddress: deviceInfo.IPAddress,
		ExpiresAt: now.Add(s.cfg.JWT.RefreshExpiry),
	}

	if deviceInfo.UserAgent != "" {
		refreshTokenEntity.Browser = &deviceInfo.UserAgent
	}

	if err := s.refreshTokenRepo.Create(ctx, refreshTokenEntity); err != nil {
		return nil, errors.Wrap(err, "TOKEN_STORAGE_FAILED", "Failed to store refresh token", 500)
	}

	// Set active session in Redis
	sessionData := &cache.SessionData{
		RefreshTokenID: refreshJti,
		IPAddress:      deviceInfo.IPAddress,
		DeviceInfo:     deviceInfo.UserAgent,
		CreatedAt:      now.Format(time.RFC3339),
	}
	if err := s.cache.SetActiveSession(ctx, userID.String(), deviceInfo.DeviceID, sessionData, s.cfg.JWT.RefreshExpiry); err != nil {
		logger.Warn().Err(err).Msg("Failed to set active session in cache")
	}

	return &AuthTokens{
		AccessToken:  accessTokenString,
		RefreshToken: refreshTokenString,
		ExpiresIn:    int64(s.cfg.JWT.AccessExpiry.Seconds()),
	}, nil
}

func (s *TokenService) VerifyAccessToken(ctx context.Context, tokenString string) (*AccessTokenClaims, error) {
	token, err := jwt.ParseWithClaims(tokenString, &AccessTokenClaims{}, func(token *jwt.Token) (interface{}, error) {
		if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
			return nil, fmt.Errorf("unexpected signing method: %v", token.Header["alg"])
		}
		return []byte(s.cfg.JWT.AccessSecret), nil
	})

	if err != nil {
		if err == jwt.ErrTokenExpired {
			return nil, errors.ErrTokenExpired
		}
		return nil, errors.ErrTokenInvalid
	}

	claims, ok := token.Claims.(*AccessTokenClaims)
	if !ok || !token.Valid {
		return nil, errors.ErrTokenInvalid
	}

	// Check if token is blacklisted
	blacklisted, err := s.cache.IsTokenBlacklisted(ctx, claims.ID)
	if err != nil {
		logger.Warn().Err(err).Msg("Failed to check token blacklist")
	}
	if blacklisted {
		return nil, errors.ErrTokenBlacklisted
	}

	return claims, nil
}

func (s *TokenService) VerifyRefreshToken(ctx context.Context, tokenString string) (*RefreshTokenClaims, error) {
	token, err := jwt.ParseWithClaims(tokenString, &RefreshTokenClaims{}, func(token *jwt.Token) (interface{}, error) {
		if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
			return nil, fmt.Errorf("unexpected signing method: %v", token.Header["alg"])
		}
		return []byte(s.cfg.JWT.RefreshSecret), nil
	})

	if err != nil {
		if err == jwt.ErrTokenExpired {
			return nil, errors.ErrRefreshTokenExpired
		}
		return nil, errors.ErrRefreshTokenInvalid
	}

	claims, ok := token.Claims.(*RefreshTokenClaims)
	if !ok || !token.Valid {
		return nil, errors.ErrRefreshTokenInvalid
	}

	return claims, nil
}

func (s *TokenService) RotateRefreshToken(ctx context.Context, oldTokenString string, deviceInfo *DeviceInfo) (*AuthTokens, error) {
	// Verify old refresh token
	claims, err := s.VerifyRefreshToken(ctx, oldTokenString)
	if err != nil {
		return nil, err
	}

	// Find token in database
	tokenID, err := uuid.Parse(claims.ID)
	if err != nil {
		return nil, errors.ErrRefreshTokenInvalid
	}

	storedToken, err := s.refreshTokenRepo.FindByID(ctx, tokenID)
	if err != nil {
		return nil, errors.ErrRefreshTokenInvalid
	}

	// Verify token is not revoked
	if storedToken.IsRevoked {
		return nil, errors.ErrTokenRevoked
	}

	// Verify token hash matches
	if !verifyTokenHash(oldTokenString, storedToken.TokenHash) {
		return nil, errors.ErrRefreshTokenInvalid
	}

	// Revoke old token
	if err := s.refreshTokenRepo.Revoke(ctx, tokenID, "token_rotation"); err != nil {
		logger.Error().Err(err).Msg("Failed to revoke old refresh token")
	}

	// Generate new token pair
	userID, err := uuid.Parse(claims.Subject)
	if err != nil {
		return nil, errors.ErrRefreshTokenInvalid
	}

	// Use existing device ID from token if not provided
	if deviceInfo.DeviceID == "" {
		deviceInfo.DeviceID = claims.DeviceID
	}

	return s.GenerateTokenPair(ctx, userID, "", deviceInfo)
}

func (s *TokenService) RevokeToken(ctx context.Context, jti string, remainingTTL time.Duration) error {
	return s.cache.BlacklistToken(ctx, jti, remainingTTL)
}

func (s *TokenService) RevokeAllUserTokens(ctx context.Context, userID uuid.UUID) error {
	return s.refreshTokenRepo.RevokeAllByUserID(ctx, userID, "logout_all_devices")
}

func (s *TokenService) RevokeDeviceTokens(ctx context.Context, userID uuid.UUID, deviceID string) error {
	err := s.refreshTokenRepo.RevokeByDeviceID(ctx, userID, deviceID, "device_logout")
	if err != nil {
		return err
	}
	return s.cache.RemoveActiveSession(ctx, userID.String(), deviceID)
}

func (s *TokenService) GetUserSessions(ctx context.Context, userID uuid.UUID) ([]model.RefreshToken, error) {
	return s.refreshTokenRepo.FindActiveByUserID(ctx, userID)
}

func (s *TokenService) CleanupExpiredTokens(ctx context.Context) (int64, error) {
	return s.refreshTokenRepo.DeleteExpired(ctx)
}

// Helper functions
func hashToken(token string) string {
	hash := sha256.Sum256([]byte(token))
	return hex.EncodeToString(hash[:])
}

func verifyTokenHash(token, hash string) bool {
	return hashToken(token) == hash
}

func HashPassword(password string) (string, error) {
	bytes, err := bcrypt.GenerateFromPassword([]byte(password), BcryptCost)
	if err != nil {
		return "", err
	}
	return string(bytes), nil
}

func CheckPassword(password, hash string) bool {
	err := bcrypt.CompareHashAndPassword([]byte(hash), []byte(password))
	return err == nil
}
